### PR TITLE
Heal Test Suite, Orbital Convergence: OCC's Open-Heart Surgery VI

### LINKED ISSUE #1777: Level Shifting in OCC
I've been investigating orbital convergence problems in the `occ` module, such as in [this test case](https://github.com/psi4/psi4/blob/master/tests/omp2-3/input.dat). I find that convergence massively improves when I turn level shifting _off_, as it is [on by default](https://github.com/psi4/psi4/blob/master/psi4/src/read_options.cc#L2695). `grep -inr 'level_shift' .` indicates that the level shifting is currently being used exclusively [to scale the diagonal elements of the B matrix in the orbital optimization step of DIIS](https://github.com/psi4/psi4/blob/master/psi4/src/psi4/occ/diis.cc#L74-L77).

@bozkaya, I know this is old code, but can you comment on why you did the level shifting this way? This doesn't seem similar to any level shifting scheme I've ever seen. My quick numerical tests indicate that this makes convergence _much worse_, and getting rid of this heals some tests that @loriab is currently concerned about. Comments from @susilehtola are also welcome for this one.

### LINKED ISSUE #1793: Broken tests
Commits prior to PR 65c0921213e6fe2979733f6c316915f79051f558 seem to have broken some tests that I suspect are not checked by our CI system:
```
	255 - ocepa-freq1 (Failed)
	259 - ocepa2 (Failed)
	260 - ocepa3 (Failed)
	273 - omp3-2 (Failed)
	278 - omp3-grad2 (Failed)
	320 - pywrap-checkrun-rhf (Failed)
	321 - pywrap-checkrun-rohf (Failed)
	322 - pywrap-checkrun-uhf (Failed)
```

The following are known to fail due to poor convergence of the occ module
```
	260 - ocepa3 (Failed)
	263 - omp2-3 (Failed)
	274 - omp3-3 (Failed)
```

The error message for `pywrap-checkrun-rhf` is also consistent with convergence issues:
```
  45     -75.0144633211      0.00e+00       1.85e-05         1.42e-04        3.01e-15
  46     -75.0144633211      0.00e+00       1.85e-05         1.42e-04        1.73e-15
  47     -75.0144633211     -1.42e-14       1.85e-05         1.42e-04        9.95e-16
  48     -75.0144633211      0.00e+00       1.85e-05         1.42e-04        5.65e-16
  49     -75.0144633211      0.00e+00       1.85e-05         1.42e-04        3.27e-16
  50     -75.0144633211      0.00e+00       1.85e-05         1.42e-04        1.88e-16
 ======================== OCEPA IS NOT CONVERGED IN 50 ITERATIONS ============
Traceback (most recent call last):
  File "/Users/fevange/Source/psi4/objdir-Debug/stage/bin/psi4", line 331, in <module>
    exec(content)
  File "<string>", line 54, in <module>
  File "/Users/fevange/Source/psi4/objdir-Debug/stage/lib/psi4/driver/driver.py", line 711, in gradient
    wfn = procedures['gradient'][lowername](lowername, molecule=molecule, **kwargs)
  File "/Users/fevange/Source/psi4/objdir-Debug/stage/lib/psi4/driver/procrouting/proc.py", line 752, in select_olccd_gradient
    return func(name, **kwargs)
  File "/Users/fevange/Source/psi4/objdir-Debug/stage/lib/psi4/driver/procrouting/proc.py", line 2147, in run_occ_gradient
    occ_wfn = core.occ(ref_wfn)
RuntimeError:
Fatal Error: OCC iterations did not converge
Error occurred in file: /Users/fevange/Source/psi4/psi4/src/psi4/occ/occ_iterations.cc on line: 363
The most recent 5 function calls were:
Printing out the relevant lines from the Psithon --> Python processed input file:
    for method in Earray:
       water.reset_point_group(sym)
       water.fix_orientation(True)
       water.update_geometry()
       banner('Testing %s' % (method))
-->    G, wfn = gradient(method, return_wfn=True)
       ans = wfn.energy()
       print('%16s\t%s\t%s\t%s' % (method, water.schoenflies_symbol(), variable('CURRENT ENERGY') - ans, ans))
       clean()
    print('  Checking gradient methods (excited state) ...')
    Earray = [
    Psi4 stopped on: Friday, 31 January 2020 05:10PM
    Psi4 wall time for execution: 0:00:06.47
```

### PR DESCRIPTION
**Pull request by JonathonMisiewicz**
## Description
At long, long, long last fixes #1777 and fixes #1793. We now DIIS on the combined vector of orbital rotation amplitudes and T2 amplitudes. This is the approach taken by the orbital-optimized DCT code, which has not had nearly the problems that the `occ` module has had with orbital convergence. `occ` should now have much more robust convergence and _not_ be one of the banes of the test suite.

In addition, full tests now pass, minus things involving conda dependencies. `v2rdm_casscf` was (unsurprisingly) segfaulting, and I was getting an error with `mp2d` that seems to trace back to me having an outdated version of QCEngine.

I had to make some changes to `opt-irc-2` to get the test suite working. See #1793 for my diagnosis of the problem. In brief, the IRC in Psi can overshoot the endpoint, causing some numerical instability in the steps post-overshoot. This is a problem to fix in PyOptking, so I just changed the test value.

This concludes the OCC's Open-Heart Surgery series.

## Todos
Notable points (developer or user-interest) that this PR has or will accomplish.
- [x] Coupled orbital and amplitude DIIS for improved convergence in `occ`
- [x] Validated every `occ` test with OMP2, OMP3, OMP2.5, OCEPA that uses no spin-scaling against Psi4Numpy implementations converged with obsessive tightness

## Checklist
- [x] Full tests pass, except for conda dependencies. That means `ctest` and `pytest`.

## Status
- [x] Ready for review
- [x] Ready for merge

### CODE REVIEW

File: psi4/src/psi4/libpsio/error.cc [hunk 1]
```diff
@@ -123,6 +123,12 @@ void psio_error(size_t unit, size_t errval) {
             fprintf(stderr, "Open failed because unit %zu exceeds ", unit);
             fprintf(stderr, "PSIO_MAXUNIT = %d.\n", PSIO_MAXUNIT);
             break;
+        case PSIO_ERROR_UNOPENED:
+            fprintf(stderr, "PSIO_ERROR: %d (File not opened)\n", PSIO_ERROR_UNOPENED);
+            fprintf(stderr, "You need to open file %zu before you attempt this operation,\n", unit);
+            fprintf(stderr, "If you're an user, contact developers immediately. This is a bug.\n");
+            fprintf(stderr, "If you're a developer, get yourself some coffee.\n");
```
Comment (dgasmith) [inline psi4/src/psi4/libpsio/error.cc]: Nice.

**Comment by JonathonMisiewicz** [comment]
Anything involving tensor contractions is just moving code around, and isn't something I'm worried about. Test updates, I'm confident are correct. Things worth checking:
* Code readability
* Nothing obviously suspicious in the DIIS call or setup in `occ_iterations`
* `opt-irc-2`, are we okay punting on this until PyOptking?

It may also be worth letting @susilehtola have a look, since he has experience with OO.

Now that I type this out, it occurs to me that I may also need to change some of the DIIS keywords along with the DIIS behavior change... As I recall, UB had separate keywords for OO or non-OO DIIS.

File: psi4/src/psi4/libpsio/error.cc (hunk 1 above)
Comment (andysim) [reply psi4/src/psi4/libpsio/error.cc]: ☕️

Comment (jturney) [reply psi4/src/psi4/libpsio/error.cc]: Good idea 👍

File: psi4/src/psi4/occ/trans_ints_uhf.cc [hunk 2]
```diff
@@ -976,19 +976,19 @@ void OCCWave::denominators_uhf() {
...
     global_dpd_->file2_close(&Fd);
 
     global_dpd_->file2_init(&Fd, PSIF_LIBTRANS_DPD, 0, ID('V'), ID('V'), "FD <V|V>");
-    FockA->get_block({nalphapi_, nmopi_}, {nalphapi_, nmopi_})->write_to_dpdfile2(&Fd);
+    FockA->get_block({nalphapi_, nmopi_ - frzvpi_}, {nalphapi_, nmopi_ - frzvpi_})->write_to_dpdfile2(&Fd);
     global_dpd_->file2_close(&Fd);
 
     global_dpd_->file2_init(&Fd, PSIF_LIBTRANS_DPD, 0, ID('v'), ID('v'), "FD <v|v>");
-    FockB->get_block({nbetapi_, nmopi_}, {nbetapi_, nmopi_})->write_to_dpdfile2(&Fd);
+    FockB->get_block({nbetapi_, nmopi_ - frzvpi_}, {nbetapi_, nmopi_ - frzvpi_})->write_to_dpdfile2(&Fd);
```
Comment (loriab) [inline psi4/src/psi4/occ/trans_ints_uhf.cc]: nice to have the `freeze_core` option hooked up to code.

File: tests/dfmp2-4/input.dat [hunk 3]
```diff
@@ -123,6 +123,10 @@ compare_values(df_mp2tot, variable('CURRENT ENERGY'), 5, 'mp2 tot')
 compare_values(df_mp2tot, val, 5, 'mp2 return')                                              #TEST
 clean()
 
+# OCC has different keyword names from DFOCC. OCC's behavior should be the standard.
+set os_scale 0.5
+set ss_scale 0.5
```
Comment (loriab) [inline tests/dfmp2-4/input.dat]: I have a different plan for this but good for now.

File: tests/ocepa-freq1/input.dat [hunk 4]
```diff
@@ -13,15 +13,14 @@ set {
...
-grad_freqs.set(0, 4, 1977.857) #TEST
-grad_freqs.set(0, 5, 3406.952) #TEST
-grad_freqs.set(0, 6, 3497.387) #TEST
+grad_freqs.set(0, 0, 513.982) #TEST
+grad_freqs.set(0, 1, 513.982) #TEST
+grad_freqs.set(0, 2, 722.325) #TEST
+grad_freqs.set(0, 3, 722.325) #TEST
+grad_freqs.set(0, 4, 1977.865) #TEST
+grad_freqs.set(0, 5, 3406.975) #TEST
+grad_freqs.set(0, 6, 3497.404) #TEST
```
Comment (loriab) [inline tests/ocepa-freq1/input.dat]: For grads/freqs, it'd be handy to know where new refs are from, whether findif or p4n implementation or just plain the code at a moment when you trust it.

File: tests/omp2-2/input.dat [hunk 5]
```diff
@@ -15,7 +15,6 @@ set {
   reference rohf
   guess gwh
   mp2_type conv
-  max_mograd_convergence 3e-4
```
Comment (loriab) [inline tests/omp2-2/input.dat]: +1 for relieving the test of this hack

File: psi4/src/psi4/occ/occ_iterations.cc [hunk 6]
```diff
@@ -410,5 +451,176 @@ void OCCWave::occ_iterations() {
...
+            }
+            global_dpd_->buf4_close(&Taa);
+            global_dpd_->buf4_close(&Tab);
+            global_dpd_->buf4_close(&Tbb);
+            global_dpd_->buf4_close(&Raa);
+            global_dpd_->buf4_close(&Rab);
+            global_dpd_->buf4_close(&Rbb);
+        }
+    }
+    psio_->close(PSIF_OCC_DPD, 1);
```
Comment (loriab) [inline psi4/src/psi4/occ/occ_iterations.cc]: I have high hopes of some of these dpd closes enabling better occ testing by running one calc after another.

File: psi4/src/psi4/occ/occ_iterations.cc [hunk 7]
```diff
@@ -316,13 +347,23 @@ void OCCWave::occ_iterations() {
...
+        oo_diis();
+
+        // Handle any needed resorts of amplitudes, after the DIIS.
+        if (wfn_type_ == "OCEPA") {
+            cepa_chemist();
+        } else if (wfn_type_ == "OMP2.5" || wfn_type_ == "OMP3") {
+            iterative_mp_postdiis_amplitudes();
+        }
+
+    } while (true); // TODO: This is a very silly do while loop.
```
Comment (loriab) [inline psi4/src/psi4/occ/occ_iterations.cc]: When exit criteria get complicated, silly while loops are sound https://github.com/psi4/psi4/blob/master/psi4/driver/procrouting/scf_proc/scf_iterator.py#L267

File: tests/ocepa-freq1/input.dat (hunk 4 above)
Comment (JonathonMisiewicz) [reply tests/ocepa-freq1/input.dat]: All energies and gradients are P4N analytic, while frequencies are by finite difference of P4N analytic gradients. All have <10^-12 residuals on everything, so it's essentially machine precision. I'll update the test inputs to say so tomorrow.

FWIW, my hesitation against doing so was that it's my private P4N implementation, which won't be very useful when people are looking at this two years from now. I'll get said implementations in a P4N-acceptable form and submit them before I graduate, but there are two Very Important Papers that need to be written before I get back to the project that would be for.

Comment (loriab) [reply tests/ocepa-freq1/input.dat]: Thanks for adding notes. The origin is more a breadcrumb guide as to independence of numbers and which university filesys to search if desperate than enough detail to re-run the numbers.

And two more 👍 to this PR for fixing two classes of occ calcs that wouldn't run for me yesterday.

Comment (JonathonMisiewicz) [reply tests/ocepa-freq1/input.dat]: To be fair, those `occ` calcs weren't running because I screwed up #1829. I am looking forward to #1841.

Comment (JonathonMisiewicz) [reply tests/ocepa-freq1/input.dat]: And, done.

**Pull request merged**